coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go

```diff
@@ -0,0 +1,378 @@
+	"github.com/eclipse-symphony/symphony/coa/pkg/apis/v1alpha2/observability"
+	observ_utils "github.com/eclipse-symphony/symphony/coa/pkg/apis/v1alpha2/observability/utils"
+	providers "github.com/eclipse-symphony/symphony/coa/pkg/apis/v1alpha2/providers"
+	states "github.com/eclipse-symphony/symphony/coa/pkg/apis/v1alpha2/providers/states"
+	"github.com/eclipse-symphony/symphony/coa/pkg/logger"
+	"github.com/go-redis/redis/v7"
+)
+
+var rLog = logger.NewLogger("coa.runtime")
+
+const (
+	entryCountPerList = 100
+	separator         = ":"
+)
+
+type RedisStateProviderConfig struct {
+	Name        string `json:"name"`
+	Host        string `json:"host"`
+	Password    string `json:"password,omitempty"`
+	RequiresTLS bool   `json:"requiresTLS,omitempty"`
+}
+
+func RedisStateProviderConfigFromMap(properties map[string]string) (RedisStateProviderConfig, error) {
+	ret := RedisStateProviderConfig{}
+	if v, ok := properties["name"]; ok {
+		ret.Name = v // providers.LoadEnv(v)
+	}
+	if v, ok := properties["host"]; ok {
+		ret.Host = v //providers.LoadEnv(v)
+	} else {
+		return ret, v1alpha2.NewCOAError(nil, "Redis pub-sub provider host name is not set", v1alpha2.BadConfig)
+	}
+	if v, ok := properties["password"]; ok {
+		ret.Password = v //providers.LoadEnv(v)
+	}
+	if v, ok := properties["requiresTLS"]; ok {
+		val := v //providers.LoadEnv(v)
+		if val != "" {
+			bVal, err := strconv.ParseBool(val)
+			if err != nil {
+				return ret, v1alpha2.NewCOAError(err, "invalid bool value in the 'requiresTLS' setting of Redis pub-sub provider", v1alpha2.BadConfig)
+			}
+			ret.RequiresTLS = bVal
+		}
+	}
+	//TODO: Finish this
+	return ret, nil
+}
+
+type RedisStateProvider struct {
+	Config  RedisStateProviderConfig
+	Context *contexts.ManagerContext
+	Client  *redis.Client
+}
+
+func (r *RedisStateProvider) ID() string {
+	return r.Config.Name
+}
+
+func (r *RedisStateProvider) SetContext(ctx *contexts.ManagerContext) {
+	r.Context = ctx
+}
+
+func (i *RedisStateProvider) InitWithMap(properties map[string]string) error {
+	config, err := RedisStateProviderConfigFromMap(properties)
+	if err != nil {
+		return err
+	}
+	return i.Init(config)
+}
+
+func (r *RedisStateProvider) Init(config providers.IProviderConfig) error {
+	vConfig, err := toRedisStateProviderConfig(config)
+	if err != nil {
+		rLog.Debugf("  P (Redis State): failed to parse provider config %+v", err)
+		return v1alpha2.NewCOAError(nil, "provided config is not a valid redis pub-sub provider config", v1alpha2.BadConfig)
+	}
+	r.Config = vConfig
+	if r.Config.Host == "" {
+		return v1alpha2.NewCOAError(nil, "Redis host is not supplied", v1alpha2.MissingConfig)
+	}
+
+	options := &redis.Options{
+		Addr:            r.Config.Host,
+		Password:        r.Config.Password,
+		DB:              0,
+		MaxRetries:      3,
+		MaxRetryBackoff: time.Second * 2,
+	}
+	if r.Config.RequiresTLS {
+		options.TLSConfig = &tls.Config{
+			InsecureSkipVerify: !r.Config.RequiresTLS,
```

Comment:
I think require TLS and skipping certificate validation are two separate things - one to indicate if TLS is needed, the other indicates if certificate validation is skipped (such as when an untrusted self-issued certificate is used)?

Reply:
I copied it from Redis pubsub configuration. Maybe we can change the configuration in both places when fixing issue https://github.com/eclipse-symphony/symphony/issues/253?